File: src/server/quizContentFingerprint.ts
import { createHash } from "node:crypto";

type QuizContent = {
  tag: string;
  difficulty: string;
  sentenceKo: string;
  choices: {
    text: string;
    isCorrect: boolean;
  }[];
};

const FIELD_SEPARATOR = "\u001f";
const CHOICE_SEPARATOR = "\u001d";

export function createQuizContentFingerprint(input: QuizContent) {
  const choices = input.choices
    .map(
      (choice) =>
        `${normalizeQuizText(choice.text)}${CHOICE_SEPARATOR}${choice.isCorrect ? "1" : "0"}`,
    )
    .sort();
  const content = [
    input.tag,
    input.difficulty,
    normalizeQuizText(input.sentenceKo),
    ...choices,
  ].join(FIELD_SEPARATOR);

  return createHash("sha256").update(content, "utf8").digest("hex");
}

function normalizeQuizText(value: string) {
  return value.trim().replace(/[\t\n\f\r ]+/g, " ");
}

File: backend/src/main/kotlin/io/github/dobbylee/cherryk/domain/quiz/QuizContentFingerprint.kt
package io.github.dobbylee.cherryk.domain.quiz

import java.nio.charset.StandardCharsets
import java.security.MessageDigest

data class QuizFingerprintChoice(
    val text: String,
    val isCorrect: Boolean,
)

data class QuizFingerprintInput(
    val tag: String,
    val difficulty: String,
    val sentenceKo: String,
    val choices: List<QuizFingerprintChoice>,
)

object QuizContentFingerprint {
    private const val FIELD_SEPARATOR = "\u001f"
    private const val CHOICE_SEPARATOR = "\u001d"
    private val innerWhitespace = Regex("[\\t\\n\\u000c\\r ]+")

    fun create(input: QuizFingerprintInput): String {
        val normalizedChoices =
            input.choices
                .map { choice ->
                    "${normalize(choice.text)}$CHOICE_SEPARATOR${if (choice.isCorrect) "1" else "0"}"
                }
                .sorted()
        val content =
            listOf(
                input.tag,
                input.difficulty,
                normalize(input.sentenceKo),
                *normalizedChoices.toTypedArray(),
            ).joinToString(FIELD_SEPARATOR)

        return MessageDigest
            .getInstance("SHA-256")
            .digest(content.toByteArray(StandardCharsets.UTF_8))
            .joinToString("") { byte -> "%02x".format(byte.toInt() and 0xff) }
    }

    private fun normalize(value: String): String = value.trim().replace(innerWhitespace, " ")
}
